### DISCUSSION TITLE
PyTorch objects / Notebooks

### DISCUSSION THREAD

**Discussion by gstech99** [mrdbourke/pytorch-deep-learning#747] [General]
Chapter 0, part 17 Tensor datatypes. 
You haven't mentioned objects and assignment operator yet. 
I'm learning both PyTorch and Notebooks at the same time. 
It seems odd not to assign the operation result to a new object.  

`float_16_tensor = float_32_tensor.type(torch.float16)`
creates a new object. If that object is then typed by itself, the contents are displayed.  

`float_16_tensor * float_32_tensor`
does not create a new object but displays the result. 
In normal development, wouldn't I be assigning that result to a new object? 

Does a Notebook cause bad habits of not keeping track of objects?
Perhaps the Notebook idea is, "No need to assign the result to a new object unless it will be referenced soon."
and "Keep the number of variable names to a minimum." 
If someone is learning Pytorch but not in a Notebook, I assume they need to assign first then print(new object). 
Notebooks are excellent for teaching, but aren't actual Pytorch applications developed with a "stand-alone" Pytorch?
Is that covered in a later chapter?

**Comment by shisirkha** [comment]
Hmm many people have many ways of doing the same problem.

your way is quite different.
but still, it works.

**Reply by gstech99** [reply]
Basically it's a question of Notebooks like Colab. They're great for the classroom or trying a new piece of software. 
But when I start a new application, do I not need [Pytorch] in a Python project to eventually make an executable?  
And there, I need to use print commands to display objects, not just type the object name itself, e.g `float_16_tensor`

**Reply by shisirkha** [reply]
Oh Yeah yeah, now i understand what you mean. in Jupyter notebook vs python raw code

they are different.
Jupyter notebooks allows us to save the variable data in memory, so we don't lose the variable data after the code ends also jupyter notebooks allows us to just type the variable names to print their data in a way. 
in raw python we don't get that fancy variable name only
instead, we need to use `print()` function to print anything that the variable returns.(NOTE: IN here returns means raw text not fancy graphs for that we need to use matplotlib.pyplot functions also panda functions for data Frame in pandas)

So now you understand that jupyter notebooks uses python but in a different way then raw python files (aka .py files).

I'm pretty impressed for your curiosity to ask this question and get active in getting the results.

Now if you are clear that jupyter notebooks and raw python are slightly different reply to me that you understand the difference else ask me i will try my best to help you.
(Also remember Colab notebooks are Jupiter notebooks that are modified to match their style in colab).
(You can call me shisir also no problem brother)
(also declare this as an answer if you understand).

If you want to learn more see this reddit post https://www.reddit.com/r/Python/comments/8a1u0e/what_is_jupyter_and_whats_the_difference_between/?rdt=37501
BEST REGARDS FROM:
-shisirkhan

**Reply by shisirkha** [reply]
Also don't be afraid to talk with me like Iam your friend.

I am willing to help anyone. (if I know the answer of it).

**Comment by gstech99** [comment]
Shisir,

I appreciate your reply. I do not have a problem; I was only trying to get
confirmation of my observations on notebooks (Jupyter or Colab) and to
document it for others. Many (not all) topics of software can be
demonstrated easily so notebooks are good for a learning environment. But
if I want to make an application, I cannot do it in a notebook. I need to
make a python project somewhere (e.g. PyCharm) and include packages (e.g.
PyTorch). In that project environment, I need to manage objects (assigning
names for future reference) and cannot easily display certain objects just
by stating them (as done in a notebook) but by using print(). IMO students
first learning notebooks should be made aware of these differences.

**Reply by shisirkha** [reply]
Yeah, you're right that we cannot build fully deployed and optimized model in jupyter notebooks or in colab(which is google's copy of jupyter notebook THAT ONLY SUPPORTS PYTHON 😁). instead, we need to use the default python scripts (aka .py files).

But still jupyter notebooks are great for many and many (Not all but quite much) CS (computer science) work.
(Also remember that converting a ML project to an exe is not recommended until you fit your ML project in a program)
Also, jupyter notebooks are great for plotting data, using pandas Data Frame etc.

From my personal experience of python I can say that jupyter notebooks are quite a big improvement for computer science and for not large ML projects.


But I still agree with you that using notebooks should be aware of these non-default python behaviors.

@gstech99 Btw can I get your contact in discord? we can be friends cause your question really impressed me. (If you don't mind contact me on discord.
else no problem I would appreciate that.)
ID= shisirkhan

**Comment by gstech99** [comment]
Shisir,

Thanks for your comments. If I have other questions, I’ll certainly
remember you. I appreciate the author setting up this separate place to ask
questions about his videos instead of using discord. Although I see a place
for discord (like notebooks) I’m not active there. Kind regards.

**Reply by shisirkha** [reply]
Ok if you have any questions and you can't find me just send friend request to me at ID: shsirkhan username

so if i am not that active in here i can still solve some questions of you.

Btw Iam always eager to help.😊

**Comment by gstech99** [comment]
This is a follow-up comment. Another complication of Colab is that it runs a Linux OS; nobody mentions that. I'm on a Windows machine and don't know Linux. This adds another [unexpected ] complexity to my learning curve when I need to work with files. One has to learn Files in Colab and Google Drive -- how are these related? How do I extract from a zip file? How do I save files to my C drive? Already know how to do these in Windows.

**Reply by shisirkha** [reply]
Yeah, it is possible to do these all-in windows. by using WSL, you can learn Linux in windows.

**Reply by shisirkha** [reply]
Generally speaking, you can use windows GUI ways to do these all things.

also, I separated the full food101 dataset into 34 batches of folder containing test,train folders for test and train data for incremental learning.

so, you know that if you try hard you can do anything.

**Reply by shisirkha** [reply]
also if you want to save the notebooks to your C drive or main drive you can click on file option in colab and click on export as .py or .ipynb format

**Reply by gstech99** [reply]
My point is, I DON'T WANT TO LEARN LINUX at this point. Why don't Colab tutorials tell you upfront it is Linux? I have nothing against Linux; it's just a distraction when I'm trying to learn other things. I've switched to learning Tensorflow in Visual Code, where I can utilize my Windows organization when files are needed.
